ui/plugin-system/src/components/MultiQueryEditor/QueryEditorContainer.tsx

```diff
@@ -54,44 +71,52 @@ export const TimeSeriesQueryInput = ({
           <DeleteIcon />
         </IconButton>
       </Stack>
-      {!isCollapsed && <QueryEditor value={query} onChange={(next) => onChange(index, next)} />}
+      {!isCollapsed && <QueryEditor queryTypes={queryTypes} value={query} onChange={(next) => onChange(index, next)} />}
     </Stack>
   );
 };
 
 // Props on MUI Box that we don't want people to pass because we're either redefining them or providing them in
 // this component
 type OmittedMuiProps = 'children' | 'value' | 'onChange';
-
 interface QueryEditorProps extends Omit<BoxProps, OmittedMuiProps> {
-  value: TimeSeriesQueryDefinition;
-  onChange: (next: TimeSeriesQueryDefinition) => void;
+  queryTypes: QueryPluginType[];
+  value: QueryDefinition;
+  onChange: (next: QueryDefinition) => void;
 }
 
 /**
- * Displays an editor for TimeSeriesQueryDefinition objects.
+ * Editor for a query definition. This component is responsible for rendering the plugin editor for the given query.
+ * This will allow user to select a plugin extending from the given supported query types, and then edit the plugin
+ * spec for this plugin.
+ * @param props
+ * @constructor
  */
 function QueryEditor(props: QueryEditorProps) {
-  const { value, onChange, ...others } = props;
-  const {
-    spec: { plugin },
-  } = value;
+  const { value, onChange, queryTypes, ...others } = props;
 
   const handlePluginChange: PluginEditorProps['onChange'] = (next) => {
     onChange(
       produce(value, (draft) => {
-        draft.spec.plugin = next;
+        draft.kind = next.selection.type as QueryPluginType;
```

Comment:
It miight be possible to not need the `as QueryPluginType` assertion if you inline this below instead of a separate handler function

Reply:
I'm realizing that it's not needed at all. I'm not sure why I did it in the first place but I just remove it :)
Thanks for the trick though. I mark it as resolved